File: leetsync/2300-construct-string-with-repeat-limit/construct-string-with-repeat-limit.java
class Solution {
    public String repeatLimitedString(String s, int repeatLimit) {
        int[] count = new int[26]; // Frequency array to store character counts
        // Count the frequency of each character
        for (char ch : s.toCharArray()) {// T.C: O(n)
            count[ch - 'a']++;
        }
        // Max heap (priority queue) to store characters in descending order
        PriorityQueue<Character> pq = new PriorityQueue<>((a, b) -> b - a);
        for (int i = 0; i < 26; i++) {
            if (count[i] > 0) {
                pq.add((char) ('a' + i));
            }
        }
        StringBuilder result = new StringBuilder();
        while (!pq.isEmpty()) {
            char ch = pq.poll(); // Get the largest character
            int freq = Math.min(count[ch - 'a'], repeatLimit); // Append up to 'repeatLimit' times
            for (int i = 0; i < freq; i++) {
                result.append(ch);
            }
            count[ch - 'a'] -= freq;
            // If the current character still has remaining occurrences
            if (count[ch - 'a'] > 0 && !pq.isEmpty()) {
                char nextChar = pq.poll(); // Get the next largest character
                result.append(nextChar);
                count[nextChar - 'a']--;
                // Reinsert characters into the priority queue if they still have remaining
                // occurrences
                if (count[nextChar - 'a'] > 0) {
                    pq.add(nextChar);
                }
                pq.add(ch);
            }
        }
        return result.toString();
    }
}

File: leetsync/2300-construct-string-with-repeat-limit/construct-string-with-repeat-limit.go
type pair struct {
	char  byte
	count int
}

type maxHeap []pair

func (h maxHeap) Len() int {
	return len(h)
}

func (h maxHeap) Less(i, j int) bool {
	return h[i].char > h[j].char
}

func (h maxHeap) Swap(i, j int) {
	h[i], h[j] = h[j], h[i]
}

func (h *maxHeap) Pop() any {
	old := *h
	n := len(old)
	item := old[n-1]
	*h = old[:n-1]
	return item
}

func (h *maxHeap) Push(x any) {
	*h = append(*h, x.(pair))
}

func repeatLimitedString(s string, repeatLimit int) string {
	result := []byte{}
	freq := map[byte]int{}

	for i := 0; i < len(s); i++ {
		freq[s[i]]++
	}

	h := maxHeap{}
	for char, count := range freq {
		h = append(h, pair{char: char, count: count})
	}

	heap.Init(&h)

	for len(h) > 0 {
		currentPair := heap.Pop(&h).(pair)
		k := min(currentPair.count, repeatLimit)

		for i := 0; i < k; i++ {
			result = append(result, currentPair.char)
		}

		currentPair.count -= k

		if currentPair.count > 0 {
			if len(h) == 0 {
				break
			}
			nextPair := heap.Pop(&h).(pair)
			result = append(result, nextPair.char)
			nextPair.count--
			if nextPair.count > 0 {
				heap.Push(&h, nextPair)
			}
			heap.Push(&h, currentPair)
		}
	}
	return string(result)
}